### DISCUSSION TITLE
float precision / round() precision

### DISCUSSION THREAD

**Discussion by MichaSU99** [micropython/micropython#13477] [ESP32] [answered]
Hi everybody,

I'm using a ESP32 DevKit C V2

I have some issues with the precision of floatingpoint calculations.
Here an example:

```
>>> 9487.26 + 0.01
9487.269
```

```
>>> round(9487.26 + 0.01, 2)
9487.269
```

I would expect that, at least with the round() function, a correct result is achieved.
Or is this normal behavior?

Thanks in advance, Michael.

**Comment by robert-hh** [comment]
1. It could be a problem of print
2. Single precision float has a little bit less than 7 significant digits. The number may be at the limit.

**Reply by robert-hh** [reply]
Actually it works fine here:
```
>>> 9487.26 + 0.01
9487.27
```
You could try to update the firmware. The print mechanism for floating point numbers has been updated.

**Reply by MichaSU99** [reply]
Yeah, this works.
Now using: 
MicroPython v1.22.1 on 2024-01-05; Generic ESP32 module with ESP32

Thanks!

**Reply by MichaSU99** [reply]
Hi robert,
I'm sorry for inconvenience, but there must still be a problem(?)
```9487.26 + 0.01``` now works, but:

```
i:float = 9486.0

for n in range(0, 100):
    i+=0.01
    # print(i)
    
print(i)
```
Result is **9486.977** instead of **9487.0**

If you print out the interim values, the inaccuracy starts at:
```
>>> 9486.02 + 0.01
9486.029
```
Merely in combination with ```round(i + 0.01, 2)``` I will get a correct answer...

regards, Michael.

**Reply by robert-hh** [reply, marked as answer]
I assume it's because of the fact, that there is no precise binary representation of 0.01. If you add that over and over, the error will sum up. Rounding after each step prevents this accumulation. If you write e.g.:
```
i:float = 9486.0

for n in range(0, 128):
    i += 1/128 
    # print(i)
    
print(i)
```
You will get a proper result. 1/128 can be exactly represented as binary number.

**Reply by MichaSU99** [reply]
Ok, many thanks for your explanation. ```round()``` is a appropriate workaround...

**Comment by kjm1102** [comment]
One of the things that has always puzzled me is that while the round function in upython can't seem to do the round up
```
>>> round(9487.26+.01, 2)
9487.269
```
whatever it is that does stringifying can
```
>>> '%.2f'%(9487.26+.01)
'9487.27'
```

**Reply by robert-hh** [reply]
Please update your ESP32 firmware. With a recent firmware I get:
```
MicroPython v1.23.0-preview.47.g16c6bc47c on 2024-01-17; Generic ESP32 module with ESP32
Type "help()" for more information.
>>> round(9487.26+.01, 2)
9487.27
```
